src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,237 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+
+namespace Microsoft.CodeAnalysis.FileHeaders
+{
+    internal abstract class AbstractFileHeaderCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds { get; }
+            = ImmutableArray.Create(IDEDiagnosticIds.FileHeaderMismatch);
+
+        public override FixAllProvider GetFixAllProvider()
+            => new FixAll(this);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MyCodeAction(cancellationToken => GetTransformedDocumentAsync(context.Document, cancellationToken)),
+                    diagnostic);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected abstract AbstractFileHeaderHelper FileHeaderHelper { get; }
+
+        protected abstract SyntaxTrivia EndOfLine(string text);
+        protected abstract SyntaxTriviaList ParseLeadingTrivia(string text);
+
+        private async Task<Document> GetTransformedDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            return document.WithSyntaxRoot(await GetTransformedSyntaxRootAsync(document, cancellationToken).ConfigureAwait(false));
+        }
+
+        private async Task<SyntaxNode> GetTransformedSyntaxRootAsync(Document document, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var options = document.Project.AnalyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
+            if (!options.TryGetEditorConfigOption(CodeStyleOptions.FileHeaderTemplate, out var fileHeaderTemplate)
+                || string.IsNullOrEmpty(fileHeaderTemplate))
+            {
+                // Avoid making changes if we fail to read the expected header
+                return root;
```

Comment:
is this not some sort of bug?  how would we want to fix the document unless this succeeded during analysis?

Reply:
It would be a bug if we hit this.

Reply:
Let's blow up then :)  i'm not a fan of silently papering over bugs.

Reply:
Can you be more specific? It seems like the right thing for a code fix to do if it can't do the right thing is to do nothing at all.

Reply:
I'd prefer we not check for bugs and just allow the code to blowup/crash/assert/watson if we care about encountering these things that should be impossible.  I'm not a fan of us scattering code into our features that should not ever be hit and which represent supposedly impossible cases.

Reply:
this comment still applies.  this would indicate some sort of serious bug in our underlying system IMO.  i would prefer we blow up intstead of masking this.  or, if we mask this, we should report some issue.

note that this behavior is actively bad IMO because it might lead to users not actually having anything fixed, with no way for them (or us) to know why that was.

Reply:
Now fixed